From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: srushti-patl <srpatel@equinix.com>
Date: Tue, 31 Oct 2023 11:55:23 -0700
Subject: [PATCH 04/51] fix: Updating Fabric Connection Schema for
 aside.access_point


diff --git a/docs/resources/equinix_fabric_connection.md b/docs/resources/equinix_fabric_connection.md
index edf9fc8..16548cd 100644
--- a/docs/resources/equinix_fabric_connection.md
+++ b/docs/resources/equinix_fabric_connection.md
@@ -103,6 +103,12 @@ Read-Only:
 
 - `id` (String) id
 
+<a id="nestedblock--a_side--access_point--netwrok"></a>
+### Nested Schema for `a_side.access_point.network`
+
+Required:
+- `uuid` (String) Equinix-assigned network identifier
+
 
 <a id="nestedblock--a_side--access_point--link_protocol"></a>
 ### Nested Schema for `a_side.access_point.link_protocol`
@@ -192,10 +198,6 @@ Optional:
 - `type` (String) Virtual Device type
 - `uuid` (String) Equinix-assigned Virtual Device identifier
 
-Read-Only:
-
-- `href` (String) Unique Resource Identifier
-
 
 <a id="nestedatt--a_side--access_point--account"></a>
 ### Nested Schema for `a_side.access_point.account`
diff --git a/equinix/fabric_connection_schema.go b/equinix/fabric_connection_schema.go
index a92e866..4b841d3 100755
--- a/equinix/fabric_connection_schema.go
+++ b/equinix/fabric_connection_schema.go
@@ -229,6 +229,11 @@ func createAccessPointVirtualDeviceSch() map[string]*schema.Schema {
 			Optional:    true,
 			Description: "Virtual Device type",
 		},
+		"name": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Customer-assigned Virtual Device Name",
+		},
 	}
 }
 
@@ -240,7 +245,7 @@ func createAccessPointInterface() map[string]*schema.Schema {
 			Description: "Equinix-assigned interface identifier",
 		},
 		"id": {
-			Type:        schema.TypeString,
+			Type:        schema.TypeInt,
 			Computed:    true,
 			Description: "id",
 		},
diff --git a/equinix/fabric_mapping_helper.go b/equinix/fabric_mapping_helper.go
index 5dd709a..e8febd2 100644
--- a/equinix/fabric_mapping_helper.go
+++ b/equinix/fabric_mapping_helper.go
@@ -249,7 +249,6 @@ func simplifiedServiceProfileToFabric(profileList []interface{}) v4.SimplifiedSe
 
 func locationToFabric(locationList []interface{}) v4.SimplifiedLocation {
 	sl := v4.SimplifiedLocation{}
-	log.Printf("[DEBUG]value of sl", sl)
 	for _, ll := range locationList {
 		llMap := ll.(map[string]interface{})
 		metroName := llMap["metro_name"]
