From bdcb620fef072f3c8713ff0d392fe23667d9ec3f Mon Sep 17 00:00:00 2001
From: ocobleseqx <oscar.cobles@eu.equinix.com>
Date: Thu, 16 Nov 2023 18:47:17 +0100
Subject: [PATCH] user agent

Signed-off-by: ocobleseqx <oscar.cobles@eu.equinix.com>
---
 internal/config/config.go | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/internal/config/config.go b/internal/config/config.go
index f16ddc7..6795459 100644
--- a/internal/config/config.go
+++ b/internal/config/config.go
@@ -23,7 +23,6 @@ import (
 	"github.com/hashicorp/go-retryablehttp"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/meta"
 	"github.com/packethost/packngo"
 	xoauth2 "golang.org/x/oauth2"
 )
@@ -277,9 +276,8 @@ func MetalRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool
 	return false, nil
 }
 
-func terraformUserAgent(version string) string {
-	ua := fmt.Sprintf("HashiCorp Terraform/%s (+https://www.terraform.io) Terraform Plugin SDK/%s",
-		version, meta.SDKVersionString())
+func pulumiUserAgent(version string) string {
+	ua := fmt.Sprintf("Pulumi/%s (+https://www.pulumi.com)", version)
 
 	if add := os.Getenv(uaEnvVar); add != "" {
 		add = strings.TrimSpace(add)
@@ -334,7 +332,9 @@ func generateModuleUserAgentString(d *schema.ResourceData, baseUserAgent string)
 }
 
 func (c *Config) fullUserAgent(suffix string) string {
-	tfUserAgent := terraformUserAgent(c.TerraformVersion)
-	userAgent := fmt.Sprintf("%s terraform-provider-equinix/%s %s", tfUserAgent, version.ProviderVersion, suffix)
+	// FORK: this gives us the ability to add a Pulumi Specific user agent
+	// TODO (ocobles) override pulumi sdk version with ldflags
+	pulumiUserAgent := pulumiUserAgent("3.0")
+	userAgent := fmt.Sprintf("%s pulumi-equinix/%s %s", pulumiUserAgent, version.ProviderVersion, suffix)
 	return strings.TrimSpace(userAgent)
 }
