From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Charles Treatman <ctreatman@equinix.com>
Date: Thu, 14 Dec 2023 11:01:47 -0600
Subject: [PATCH 49/51] refactor: switch from metal-go to equinix-sdk-go (#489)


diff --git a/equinix/data_source_metal_device.go b/equinix/data_source_metal_device.go
index 8e7e9ab..ed5548b 100644
--- a/equinix/data_source_metal_device.go
+++ b/equinix/data_source_metal_device.go
@@ -10,7 +10,7 @@ import (
 
 	"github.com/equinix/terraform-provider-equinix/internal/config"
 
-	metalv1 "github.com/equinix-labs/metal-go/metal/v1"
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
 )
diff --git a/equinix/data_source_metal_device_bgp_neighbors.go b/equinix/data_source_metal_device_bgp_neighbors.go
index 641f2f8..3adf684 100644
--- a/equinix/data_source_metal_device_bgp_neighbors.go
+++ b/equinix/data_source_metal_device_bgp_neighbors.go
@@ -5,7 +5,7 @@ import (
 
 	"github.com/equinix/terraform-provider-equinix/internal/config"
 
-	metalv1 "github.com/equinix-labs/metal-go/metal/v1"
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
diff --git a/equinix/data_source_metal_devices.go b/equinix/data_source_metal_devices.go
index 9e1d834..dee58d7 100644
--- a/equinix/data_source_metal_devices.go
+++ b/equinix/data_source_metal_devices.go
@@ -6,7 +6,7 @@ import (
 
 	"github.com/equinix/terraform-provider-equinix/internal/config"
 
-	metalv1 "github.com/equinix-labs/metal-go/metal/v1"
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/equinix/terraform-provider-equinix/internal/datalist"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
diff --git a/equinix/helpers_device.go b/equinix/helpers_device.go
index 74ecceb..0a2932c 100644
--- a/equinix/helpers_device.go
+++ b/equinix/helpers_device.go
@@ -16,7 +16,7 @@ import (
 
 	equinix_errors "github.com/equinix/terraform-provider-equinix/internal/errors"
 
-	metalv1 "github.com/equinix-labs/metal-go/metal/v1"
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
diff --git a/go.mod b/go.mod
index c6f09fb..e14e4ce 100644
--- a/go.mod
+++ b/go.mod
@@ -5,8 +5,8 @@ go 1.20
 require (
 	github.com/antihax/optional v1.0.0
 	github.com/equinix-labs/fabric-go v0.7.1
-	github.com/equinix-labs/metal-go v0.29.0
 	github.com/equinix/ecx-go/v2 v2.3.1
+	github.com/equinix/equinix-sdk-go v0.30.0
 	github.com/equinix/ne-go v1.11.0
 	github.com/equinix/oauth2-go v1.0.0
 	github.com/equinix/rest-go v1.3.0
diff --git a/go.sum b/go.sum
index 186af92..6ac881b 100644
--- a/go.sum
+++ b/go.sum
@@ -261,12 +261,10 @@ github.com/envoyproxy/go-control-plane v0.10.2-0.20220325020618-49ff273808a1/go.
 github.com/envoyproxy/protoc-gen-validate v0.1.0/go.mod h1:iSmxcyjqTsJpI2R4NaDN7+kN2VEUnK/pcBlmesArF7c=
 github.com/equinix-labs/fabric-go v0.7.1 h1:4yk0IKXMcc72rkRVbcYHokAEc1uUB06t6NXK+DtSsbs=
 github.com/equinix-labs/fabric-go v0.7.1/go.mod h1:oqgGS3GOI8hHGPJKsAwDOEX0qRHl52sJGvwA/zMSd90=
-github.com/equinix-labs/metal-go v0.27.0 h1:p5Bqus/gSs5oQezHWXWpc0IzkQl06+yZgbXT5jB7AWs=
-github.com/equinix-labs/metal-go v0.27.0/go.mod h1:SmxCklxW+KjmBLVMdEXgtFO5gD5/b4N0VxcNgUYbOH4=
-github.com/equinix-labs/metal-go v0.29.0 h1:dTi48+ni4Xbey0VJjHrkoAOxJvVPl9ukA/6DoJv6o8s=
-github.com/equinix-labs/metal-go v0.29.0/go.mod h1:SmxCklxW+KjmBLVMdEXgtFO5gD5/b4N0VxcNgUYbOH4=
 github.com/equinix/ecx-go/v2 v2.3.1 h1:gFcAIeyaEUw7S8ebqApmT7E/S7pC7Ac3wgScp89fkPU=
 github.com/equinix/ecx-go/v2 v2.3.1/go.mod h1:FvCdZ3jXU8Z4CPKig2DT+4J2HdwgRK17pIcznM7RXyk=
+github.com/equinix/equinix-sdk-go v0.30.0 h1:u/+/p00mfAhDhoLvP1jTKruXndAYWoTwqN65BTbAPCg=
+github.com/equinix/equinix-sdk-go v0.30.0/go.mod h1:qnpdRzVftHFNaJFk1VSIrAOTLrIoeDrxzUr3l8ARyvQ=
 github.com/equinix/ne-go v1.11.0 h1:ja6G2fmcGrLsOeV25Mq6pDfH+/cUlvxJbnE8uRXTGGk=
 github.com/equinix/ne-go v1.11.0/go.mod h1:eHkkxM4nbTB7DZ9X9zGnwfYnxIJWIsU3aHA+FAoZ1EI=
 github.com/equinix/oauth2-go v1.0.0 h1:fHtAPGq82PdgtK5vEThs8Vwz6f7D/8SX4tE3NJu+KcU=
diff --git a/internal/config/config.go b/internal/config/config.go
index d5d116b..7badd59 100644
--- a/internal/config/config.go
+++ b/internal/config/config.go
@@ -15,8 +15,8 @@ import (
 	"time"
 
 	v4 "github.com/equinix-labs/fabric-go/fabric/v4"
-	metalv1 "github.com/equinix-labs/metal-go/metal/v1"
 	"github.com/equinix/ecx-go/v2"
+	"github.com/equinix/equinix-sdk-go/services/metalv1"
 	"github.com/equinix/ne-go"
 	"github.com/equinix/oauth2-go"
 	"github.com/equinix/terraform-provider-equinix/version"
